### ISSUE TITLE
Add support for cuda graph with torch.cond

### ISSUE THREAD

**Issue by galv** [pytorch/pytorch#168911]
### 🚀 The feature, motivation and pitch

Control flow is typically banned during stream capture to a cuda graph. This is banned because control flow like this will implicitly call item() on `x.sum() > 0` to copy the boolean condition from device to host, which does cudaStreamSynchronize():

```
# suppose x is a tensor on a cuda device
if x.sum() > 0:
    return x
else
    return torch.zeros_like(x)
```

Thus, control flow today requires the usage of torch.compile with cudagraph trees instead.

However, it turns out that torch.cond() composes nicely with CUDA Graph conditional nodes, because the arguments to torch.cond() map very nicely to [the parameters required](https://github.com/NVIDIA/cuda-samples/blob/c94ff366aed18c797b8a85dfaac7817b0228b420/Samples/3_CUDA_Features/graphConditionalNodes/graphConditionalNodes.cu#L102-L108) to construct an IF cuda graph conditional node. torch.cond() takes in functions to execute for both sides of the conditional as direct arguments, allowing us to construct the "child" cuda graphs.

Here is a TODO list of items:

1. Support cuda graph capture of torch.cond() in eager mode pytorch
2. Support cuda graph capture of torch.cond() with the cudagraphs backend (not really necessary, but easy enough to add) and the "eager" backend.
3. Support cuda graph capture of torch.cond() with the inductor backend.
4. Make sure that autograd works as well.
5. Use the ELSE statement supported since [CUDA 12.8](https://developer.nvidia.com/blog/dynamic-control-flow-in-cuda-graphs-with-conditional-nodes/#conditional_nodes), rather than two IF nodes to implement this, for a small performance boost. Note that pytorch supports only CUDA12.6 and newer right now, so I might require CUDA 12.8 or higher for this feature for simplicity. We will see.

There was an attempt to implement this over a year ago here: https://github.com/pytorch/pytorch/pull/140979/

The biggest problem back then was that torch DispatchModes did not support higher order ops, but this was added recently: https://github.com/pytorch/pytorch/blob/a3cc252e03572835c15afde54b81fc5e8616ad27/torch/utils/_python_dispatch.py#L88 Thank you @ydwu4 

At this point, I actually rebased that PR #140979 already at https://github.com/pytorch/pytorch/tree/galv/cudagraphs-conditional-nodes-4, just to check if Yidi's change fixed my problems, and it does. So there are not likely to be any serious issues at this point with implementing points (1), (2) and (4). I will leave the implementation of (3) as a TODO item for now because I am not familiar with the inductor backend in pytorch.

Some motivation for this: 

Firstly, the desire to do basic data-dependent computation has come up several times, but one of the key use cases is skipping the optimizer in the case of a NaN gradient in mixed precision training.

Secondly, figuring out making torch.cond() work with cuda graphs paves the way for adding support for torch.while_loop() in cuda graphs, which may be used to do MoE inside a cuda graph without dropping tokens, but that is more complicated and requires its own issue to write down those issues (they go beyond just while_loop support for cuda graphs; it may also need tricky memory management for autograd).

FYI @eellison @BoyuanFeng @tingyangk @ezyang @eee4017 @ydwu4 

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang @ydwu4 @bdhirsh

**Comment by ezyang** [comment]
We need a mutable cond. Most of the complications is around autograd which can be bypassed in inference which is what optimizer update would be. I think it is feasible, but also functionalization would try to functionalize the cond, so we need to make sure it defunctionalizes ok
